### PR TITLE
Retry txn watcher on error instead of exiting

### PR DESCRIPTION
**Pull request by wallyworld**
This PR attempts to make the txn watcher more robust. There's a suggestion there could be session contamination so the PR does:
- creates a separate session just for the txn watcher worker
- adds exponential backoff / retry if there's a mgo error; attempt to recover before reporting an error
- logs the actual txn watcher error message if the hub watcher does exit

It seems under high load, the txn watcher session gets disconnected and/or closed by mistake. So the retry will attempt to refresh the session and start the worker loop again. It will eventally report an error to the hubwatcher, which will restart all workers as occurs now on the very first error.

## QA steps

bootstrap
deploy mariadb and mediawiki
ssh to the controller
`echo "sync-error"  > /var/lib/juju/wrench/txnwatcher`
check the controller logs for txn watcher errors but the workers stay running
remove  /var/lib/juju/wrench/txnwatcher
relate the 2 charms
check for normal operations

## Bug reference

https://bugs.launchpad.net/juju/+bug/1890167

### CODE REVIEW

**Comment by petevg** [comment]
I think that I understand where you're going here. I'm a little bit concerned that we're treating a symptom, rather than a root cause, though.

Are there other points where we might be opening duplicate sessions on MongoDB? Is there some sort of governor or other structure we need to drop into place to prevent that? 

Or is the fix really just this targeted and simple?

**Review by jameinel — COMMENTED** [review]
I think this is a case where we know that we do want a separate session for txn watcher, and the original question is why the session was shared in the first place. (*Especially* for txn watcher, since it should have dedicated interaction with Mongo, as it drives most of what we do.)
I would hope that we could be careful (maybe with a strict mode?) so that this sort of retry is getting attention.

Also, in discussing this with you, we want to watch out for the case where the last-known-transaction is no longer in the log, which means we really do need to restart the world, because we don't know what changes have happened that we aren't sending events for.

File: state/watcher/txnwatcher.go [hunk 1]
```diff
@@ -50,6 +54,17 @@ var (
...
 
+	// ErrorStrategy is used to determine how long
+	// to delay between poll intervals when attempting
+	// to recover from a mongo error.
+	//
+	// It must not be changed when any watchers are active.
+	ErrorStrategy retry.Strategy = retry.Exponential{
+		Initial:  txnWatcherErrorShortWait,
+		Factor:   2.0,
+		MaxDelay: 30 * time.Second,
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/watcher/txnwatcher.go:65]: For the first cut, I wanted to be conservative about not retrying enough, to get a feel for 1. if retries work, and 2. at what point the retry succeeds. We can then wind back the retry count/backoff delay. But if we think what's there is too long, I can reduce it.

**Comment by jameinel** [comment]
We talked about this, the other thing that we are missing is if the txn sync loop doesn't find the 'lastId' then it has lost track of what is going on, and should raise a fatal error.

File: state/watcher/txnwatcher.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/watcher/txnwatcher.go:66]: I've reduced the retry count to 8 as max 2 x 30s should be enough

**Comment by wallyworld** [comment]
Added a commit to introduce an out of sync error which exits immediately

**Comment by wallyworld** [comment]
> I think that I understand where you're going here. I'm a little bit concerned that we're treating a symptom, rather than a root cause, though.
> 
> Are there other points where we might be opening duplicate sessions on MongoDB? Is there some sort of governor or other structure we need to drop into place to prevent that?
> 
> Or is the fix really just this targeted and simple?

The fix is simple and targetted. The txn watcher getting its own session instead of sharing one with the controller model state allows us to decouple those 2 functional pieces to try and tease out if there's an unexpected Close() on the controller state session which is messing with the txn watcher. We don't yet know exactly what's happening with the session close issues.

**Comment by wallyworld** [comment]
!!build!!

File: state/watcher/txnwatcher.go [hunk 2]
```diff
@@ -388,5 +472,13 @@ func (w *TxnWatcher) sync() (bool, error) {
 	if err := iter.Close(); err != nil {
 		return false, errors.Annotate(err, "watcher iteration error")
 	}
+	// If we have exited the iterator without consuming all the txns since we
+	// last synced, or the collection has looped that's an issue.
+	if lastId != nil && lastSeenId != lastId {
```
Comment (jameinel) [inline state/watcher/txnwatcher.go:477]: So this is potentially something that we want very good testing of.
AFAICT the old code was potentially very dangerous in that events could go missing. But this new code could potentially cause a lot *more* desyncs in the wild.
We should probably try to do some scale testing against it, as well as making sure it has good SolQA coverage and we are checking for these warnings/errors in the log.
This is the sort of error that I would probably like to make fatal in SolQA

**Review by jameinel — APPROVED** [review]

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**